modules/basis_universal/register_types.cpp

```diff
@@ -52,44 +54,47 @@ enum BasisDecompressFormat {
 #ifdef TOOLS_ENABLED
 static Vector<uint8_t> basis_universal_packer(const Ref<Image> &p_image, Image::UsedChannels p_channels) {
 	Vector<uint8_t> budata;
-
 	{
+		basisu::basis_compressor_params params;
 		Ref<Image> image = p_image->duplicate();
-
-		// unfortunately, basis universal does not support compressing supplied mipmaps,
-		// so for the time being, only compressing individual images will have to do.
-
-		if (image->has_mipmaps()) {
-			image->clear_mipmaps();
-		}
 		if (image->get_format() != Image::FORMAT_RGBA8) {
 			image->convert(Image::FORMAT_RGBA8);
 		}
-
-		basisu::image buimg(image->get_width(), image->get_height());
-
+		Ref<Image> image_single = image->duplicate();
 		{
-			Vector<uint8_t> vec = image->get_data();
+			if (image_single->has_mipmaps()) {
+				image_single->clear_mipmaps();
+			}
+			basisu::image buimg(image_single->get_width(), image_single->get_height());
+			Vector<uint8_t> vec = image_single->get_data();
 			const uint8_t *r = vec.ptr();
-
 			memcpy(buimg.get_ptr(), r, vec.size());
+			params.m_source_images.push_back(buimg);
 		}
-
-		basisu::basis_compressor_params params;
-		params.m_uastc = true;
-		params.m_max_endpoint_clusters = 512;
-		params.m_max_selector_clusters = 512;
+		basisu::vector<basisu::image> source_images;
+		for (int32_t mipmap_i = 1; mipmap_i < image->get_mipmap_count(); mipmap_i++) {
+			Ref<Image> mip = image->get_image_from_mipmap(mipmap_i);
+			basisu::image buimg(mip->get_width(), mip->get_height());
+			Vector<uint8_t> vec = mip->get_data();
+			const uint8_t *r = vec.ptr();
+			memcpy(buimg.get_ptr(), r, vec.size());
+			source_images.push_back(buimg);
+		}
+		params.m_source_mipmap_images.push_back(source_images);
+		params.m_quality_level = basisu::BASISU_QUALITY_MIN;
+		params.m_mip_fast = true;
+		params.m_pack_uastc_flags = basisu::cPackUASTCLevelFaster;
 		params.m_multithreading = true;
-		//params.m_quality_level = 0;
-		//params.m_disable_hierarchical_endpoint_codebooks = true;
-		//params.m_no_selector_rdo = true;
+		params.m_quality_level = 0;
```

Comment:
You're setting it twice to different values.

Reply:
Resolving.

Reply:
Thanks, you found the cause of the lower than expected performance! Thanks.